### PR TITLE
(rebased) Refactor sort/search/pagination for ClinicalData table endpoint

### PR DESCRIPTION
**Pull request by dippindots**
Fix # (see https://help.github.com/en/articles/closing-issues-using-keywords)

Describe changes proposed in this pull request:
- a
- b

# Checks
- [ ] The commit log is comprehensible. It follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). We can fix this during merge by using a squash+merge if necessary
- [ ] Has tests or has a separate issue that describes the types of test that should be created. If no test is included it should explicitly be mentioned in the PR why there is no test.
- [ ] Is this PR adding logic based on one or more **clinical** attributes? If yes, please make sure validation for this attribute is also present in the data validation / data loading layers (in backend repo) and documented in [File-Formats Clinical data section](https://github.com/cBioPortal/cbioportal/blob/master/docs/File-Formats.md#clinical-data)!
- [ ] Make sure your PR has one of the labels defined in https://github.com/cBioPortal/cbioportal/blob/master/.github/release-drafter.yml

# Any screenshots or GIFs?
If this is a new visual feature please add a before/after screenshot or gif
here with e.g. [Giphy CAPTURE](https://giphy.com/apps/giphycapture) or [Peek](https://github.com/phw/peek)

# Notify reviewers
Read our [Pull request merging
policy](../CONTRIBUTING.md#pull-request-merging-policy). It can help to figure out who worked on the
file before you. Please use `git blame <filename>` to determine that
and notify them either through slack or by assigning them as a reviewer on the PR

### CODE REVIEW

File: src/main/java/org/cbioportal/persistence/mybatis/StudyMyBatisRepository.java [hunk 1]
```diff
@@ -17,13 +17,13 @@ public class StudyMyBatisRepository implements StudyRepository {
...
     @Autowired
-    private OffsetCalculator offsetCalculator;
+    private PaginationCalculator paginationCalculator;
 
     @Override
     public List<CancerStudy> getAllStudies(String keyword, String projection, Integer pageSize, Integer pageNumber,
                                            String sortBy, String direction) {
 
-        return studyMapper.getStudies(null, keyword, projection, pageSize, offsetCalculator.calculate(pageSize, pageNumber), 
+        return studyMapper.getStudies(null, keyword, projection, pageSize, paginationCalculator.offset(pageSize, pageNumber),
```
Comment (haynescd) [inline src/main/java/org/cbioportal/persistence/mybatis/StudyMyBatisRepository.java:26]: PaginationCalculator::offset(pageSize, pageNumber)

File: src/main/java/org/cbioportal/service/impl/ClinicalDataServiceImpl.java [hunk 2]
```diff
@@ -227,17 +234,44 @@ public List<ClinicalData> getPatientClinicalDataDetailedToSample(List<String> st
...
+        List<ClinicalData> patientClinicalData = clinicalDataRepository.getPatientClinicalDataBySampleInternalIds(
+            visibleSampleInternalIds
+        );
+        
+        // Merge sample and patient level clinical data and key by unique sample-key.
+        sampleClinicalDataCollection.setByUniqueSampleKey(
+            Stream.concat(sampleClinicalData.stream(), patientClinicalData.stream())
+                .collect(Collectors.groupingBy(clinicalDatum ->
+                    calculateBase64(clinicalDatum.getSampleId(), clinicalDatum.getStudyId())
+        )));
```
Comment (haynescd) [inline src/main/java/org/cbioportal/service/impl/ClinicalDataServiceImpl.java:272]: Just from observation. Our functions are usually bigger than they need to be. We should start focusing on keeping them small. With that, I would extract this out into a fn maybe buildSampleClinicalDataCollection

File: src/main/java/org/cbioportal/utils/Encoding.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package org.cbioportal.utils;
+
+import java.util.Base64;
+
+public class Encoding {
```
Comment (haynescd) [inline src/main/java/org/cbioportal/utils/Encoding.java:5]: Encoder would be a better name

File: src/main/java/org/cbioportal/model/SampleClinicalDataCollection.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package org.cbioportal.model;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class SampleClinicalDataCollection {
```
Comment (haynescd) [inline src/main/java/org/cbioportal/model/SampleClinicalDataCollection.java:7]: Probably should start making Models not returned by MyBatis (Java Objs created by query results... Persistence layer) Immutable.

I would suggest making the class immutable (using a Builder Pattern to build the object)

**Review by haynescd — CHANGES_REQUESTED** [review]
Also some sonarCloud issues to relove

**Comment by JREastonMarks** [comment]
Weren't most, if not all, of these fixed with @fuzhaoyuan PR?

File: src/main/java/org/cbioportal/security/token/config/disable dat
Comment (haynescd) [inline src/main/java/org/cbioportal/security/token/config/disable dat:1]: These should not be added

**Comment by inodb** [comment]
Correct @JREastonMarks - I think this is superseded by: https://github.com/cBioPortal/cbioportal/pull/10407

**Pull request closed without merging**